internal/jobobject/jobobject.go

```diff
@@ -0,0 +1,414 @@
+	job.handleLock.RLock()
+	defer job.handleLock.RUnlock()
+	if job.handle == 0 {
+		return ErrAlreadyClosed
+	}
+	return windows.TerminateJobObject(job.handle, exitCode)
+}
+
+// Shutdown gracefully shuts down all the processes in the job object. Context timeout
+// will be used as timer for graceful shutdown. If all of the processes in the job have not
+// exited by the timeout, the job will be terminated.
+func (job *JobObject) Shutdown(ctx context.Context) error {
+	job.handleLock.RLock()
+	if job.handle == 0 {
+		return ErrAlreadyClosed
+	}
+	job.handleLock.RUnlock()
+
+	pids, err := job.Pids()
+	if err != nil {
+		return fmt.Errorf("failed to get pids for job object: %s", err)
+	}
+
+	for _, pid := range pids {
+		if err := windows.GenerateConsoleCtrlEvent(windows.CTRL_BREAK_EVENT, pid); err != nil {
+			log.G(ctx).WithField("pid", pid).Errorf("failed to send ctrl-break to process in job: %s", err)
+			return job.Terminate(1)
+		}
+	}
+
+	// Check if the processes have exited, it may take some time for some processes to clean up
+	// so keep checking until we've either hit the timeout set on the context or the processes
+	// clean up successfully.
+	for {
+		time.Sleep(time.Millisecond * 10)
+		// Get pids in job again to check if they were successfully killed by the signals.
+		// Certain programs handle ctrl signals so we need a fallback to actually kill them.
+		newPids, err := job.Pids()
+		if err != nil {
+			return fmt.Errorf("failed to get pids for job object: %s", err)
+		}
+		// All processes have gracefully exited.
+		if len(newPids) == 0 {
+			break
+		}
+		select {
+		case <-ctx.Done():
+			return job.Terminate(1)
+		default:
+		}
+	}
+	return nil
+}
+
+// Pids returns all of the process IDs in the job object.
+func (job *JobObject) Pids() ([]uint32, error) {
+	job.handleLock.RLock()
+	defer job.handleLock.RUnlock()
+
+	if job.handle == 0 {
+		return nil, ErrAlreadyClosed
+	}
+
+	info := winapi.JOBOBJECT_BASIC_PROCESS_ID_LIST{}
+	err := winapi.QueryInformationJobObject(
+		job.handle,
+		winapi.JobObjectBasicProcessIdList,
+		uintptr(unsafe.Pointer(&info)),
+		uint32(unsafe.Sizeof(info)),
+		nil,
+	)
+
+	// This is either the case where there is only one process or no processes in
+	// the job. Any other case will result in ERROR_MORE_DATA. Check if info.NumberOfProcessIdsInList
+	// is 1 and just return this, otherwise return an empty slice.
+	if err == nil {
+		if info.NumberOfProcessIdsInList == 1 {
+			return []uint32{uint32(info.ProcessIdList[0])}, nil
+		}
+		// Return empty slice instead of nil to play well with the caller of this.
+		// Do not return an error if no processes are running inside the job
+		return []uint32{}, nil
+	}
+
+	if err != winapi.ERROR_MORE_DATA {
+		return nil, fmt.Errorf("failed initial query for PIDs in job object: %s", err)
+	}
+
+	jobBasicProcessIDListSize := unsafe.Sizeof(info) + (unsafe.Sizeof(info.ProcessIdList[0]) * uintptr(info.NumberOfAssignedProcesses-1))
```

Comment:
Why do we use `info.NumberOfAssignedProcesses-1` here? Wouldn't that create a buffer that's not big enough to hold all of the pids?

Reply:
I confirmed this as well.